src/nbl/asset/utils/CHLSLCompiler.cpp

```diff
@@ -61,29 +92,149 @@ CHLSLCompiler::~CHLSLCompiler()
 {
-    ComPtr<IDxcBlobEncoding> errorMessages;
-    ComPtr<IDxcBlob> objectBlob;
-    ComPtr<IDxcResult> compileResult;
+    auto stageArgumentPos = std::find(arguments.begin(), arguments.end(), L"-HV");
+    if (stageArgumentPos != arguments.end()) {
+        auto index = stageArgumentPos - arguments.begin() + 1; // -HV XXXXX, get index of second
+        std::wstring version = arguments[index];
+        if (!isalpha(version.back()) && version.length() >= 4 && std::stoi(version) < 2021)
+            arguments[index] = L"2021";
+    }
+    else {
+        logger.log("Compile flag error: Required compile flag not found -HV. Force enabling -HV 202x, as it is required by Nabla.", system::ILogger::ELL_WARNING);
+        arguments.push_back(L"-HV");
+        arguments.push_back(L"202x");
+    }
+}
 
-    std::string GetErrorMessagesString()
+
+static void try_upgrade_shader_stage(std::vector<std::wstring>& arguments, asset::IShader::E_SHADER_STAGE shaderStageOverrideFromPragma, system::logger_opt_ptr& logger) {
+	
+	constexpr int MajorReqVersion = 6,
+		MinorReqVersion = 7;
+    auto overrideStageStr = ShaderStageToString(shaderStageOverrideFromPragma);
+	auto foundShaderStageArgument = std::find(arguments.begin(), arguments.end(), L"-T");
+	if (foundShaderStageArgument != arguments.end() && foundShaderStageArgument + 1 != arguments.end()) {
+		auto foundShaderStageArgumentValueIdx = foundShaderStageArgument - arguments.begin() + 1;
+		std::wstring s = arguments[foundShaderStageArgumentValueIdx];
+        if (s.length() >= 6) {
+            std::wstring stageStr, majorVersionString, minorVersionString;
+            std::vector<std::wstring::iterator> underscorePositions = {};
+            auto it = std::find(s.begin(), s.end(), '_');
+            while (it != s.end()) {
+                underscorePositions.push_back(it);
+                it = std::find(it + 1, s.end(), '_');
+            }
+
+            // Bad input check
+            if (underscorePositions.size() < 2)
+            {
+                logger.log("Incorrect -T argument value.\nExpecting string with at least 2 '_' delimiters: between shader stage, version major and version minor.",
+                    system::ILogger::ELL_ERROR);
+                return;
+            }
+
+            // Stage
+            stageStr = std::wstring(s.begin(), underscorePositions[0]);
+            if (shaderStageOverrideFromPragma != IShader::ESS_UNKNOWN) // replace first 2 characters if shaderStageOverrideFromPragma != Unknown
+            {
+                if (!overrideStageStr) {
+                    logger.log("Invalid shader stage with int value '%i'.\nThis value does not have a known string representation.",
+                        system::ILogger::ELL_ERROR, shaderStageOverrideFromPragma);
+                    return;
+                }
+                stageStr = std::wstring(overrideStageStr);
+            }
```

Comment:
why do you do the check here?

I mean you initialize the "no -T found` from this as well, so you SHOULD do this check at the top!

Reply:
it doesn't matter if you log an ERROR and don't stop the compilation beceause the override wasn't needed and the other value was ok, but the else case for find `-T` needs the check